bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/BookKeeperException.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2017 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.bookkeeper.client.api;
+
+import org.apache.bookkeeper.client.LedgerHandleAdv;
+
+/**
+ * Super class for all errors which occur using BookKeeper client
+ */
+public abstract class BookKeeperException extends Exception {
+    protected final int code;
+
+    public BookKeeperException(int code) {
+        super(getMessage(code));
+        this.code = code;
+    }
+
+    /**
+     * Get the return code for the exception
+     * @return
+     */
+    public final int getCode() {
```

Comment:
why final? I thought the code was just moved from the legacy BKException file. were you just manually type it?

Reply:
I think the adding final here is good because it would be bad for an user to overload this method

Reply:
I am fine with adding that. 

but it is always good to keep code unchanged when you moved code around. because it is hard for people to know if there was anything changed when you moved the code. when I see `final` was added, I am a bit uncomfortable on reviewing this part, because that means you changed code when moving it and I don't know how much you changed it.